[AST/Sema] NonCopyableGenerics: Address some of the request circularity issues

xedin opened the pull request:
- Extract logic related to protocol handling out of `InvertibleAnnotationRequest`
- Don't resolve requirements that don't have inverse as a constraint

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[Sema] InvertibleAnnotationRequest: Resolve requirements only if constraint is an `InverseTypeRepr`

This avoid a cycle in recursive and invalid code where one of the
requirements references the subject type.

xedin committed:
[Sema] InvertibleAnnotationRequest: Only resolve types of requirements placed on Self of protocol

xedin committed:
[AST] Add a way to check whether TypeRepr is an inverse of the given invertible protocol

xedin committed:
[AST] Simplify `ProtocolDecl::requiresInvertible` check

The method only needs to know about the presence of inverses
for the given invertiable protocol.

xedin committed:
[AST] Move `hasInverseMarking` into `ProtocolDecl` and produce a loc of found inverse

xedin committed:
[AST] InvertibleAnnotationRequest: Remove logic associated with protocol handling

kavon commented on include/swift/AST/Decl.h:
```diff
@@ -5216,6 +5216,11 @@ class ProtocolDecl final : public NominalTypeDecl {
+  /// Determine whether this protocol has ~<target>` stated on
+  /// itself, one of its inherited types or `Self` requirements.
+  std::pair</*found=*/bool, /*where=*/SourceLoc>
+  hasInverseMarking(InvertibleProtocolKind ip) const;
```
if CI fails, this would be nice to fix: name parameter `target` instead of `ip`.

xedin replied:
Done!

xedin commented:
@swift-ci please test

xedin committed:
[AST] Route inverse marking detection through `TypeDecl::hasInverseMarking`

The first step on the path to dismangle `TypeDecl::getMarking()`.

xedin committed:
[AST] Route `hasInverse` through `hasInverseMarking` and simplify it

xedin committed:
[AST] NonCopyableGenerics: Remove a few uses for `getMarking(...).getPositive()`

xedin committed:
[AST/Sema] NonCopyableGenerics: Remove all but one direct use of `getMarking(...)`

xedin commented:
@swift-ci please test

xedin committed:
[AST] NoCopyableGenerics: Remove last use of `getMarking(...)`

Introduce `NominalTypeDecl::has{Marking, InverseMarking}` methods
that split the logic from `InvertibleAnnotationRequest`

xedin committed:
[AST/Sema] NFC: Remove obsolete `getMarking(...)` and `InvertibleAnnotationRequest`

xedin commented:
@swift-ci please test

xedin committed:
[AST] NonCopyableGenerics: Don't resolve requirements in `NominalTypeDecl::hasInverseMarking`

TypeRepr for the subject should be already bound to a generic type
declaration, so there is no need to resolve the requirement to
check that.

xedin commented:
@swift-ci please test

xedin committed:
[AST] NonCopyableGenerics: Expand `hasInverseMarking` to support associated type declarations

This is required for `ASTPrinter` to wrap the protocol that has
associated type with inverses in a feature block.

xedin committed:
[Tests] NFC: Adjust previous fixed test-case

xedin committed:
[AST] Remove `TypeDecl::hasInverse` and move `canBe{Copyable, Escapable} to `NominalTypeDecl`

xedin commented:
@swift-ci please test

xedin commented:
Please test with following pull request:
https://github.com/apple/swift-docc/pull/836

@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Linux

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please clean test macOS platform

xedin commented:
@swift-ci please smoke test macOS platform
